File: app/pages/repos/_repo_id/models/_model_id/training_metrics/index/props.ts
import { ModelLayoutInfo } from 'layouts/model_layout'

export enum Type {
	Regressor = 'regressor',
	BinaryClassifier = 'binary_classifier',
	MulticlassClassifier = 'multiclass_classifier',
}

export type Props = {
	inner: Inner
	modelLayoutInfo: ModelLayoutInfo
}

export type Inner =
	| {
			type: Type.Regressor
			value: RegressorProps
	  }
	| {
			type: Type.BinaryClassifier
			value: BinaryClassifierProps
	  }
	| {
			type: Type.MulticlassClassifier
			value: MulticlassClassifierProps
	  }

export type RegressorProps = {
	baselineMse: number
	baselineRmse: number
	id: string
	mse: number
	rmse: number
}

export type BinaryClassifierProps = {
	accuracy: number
	aucRoc: number
	baselineAccuracy: number
	classes: string[]
	id: string
	precision: number
	recall: number
}

export type MulticlassClassifierProps = {
	accuracy: number
	baselineAccuracy: number
	classMetrics: Array<{
		precision: number
		recall: number
	}>
	classes: string[]
	id: string
}

File: app/pages/repos/_repo_id/models/_model_id/training_metrics/index/props.rs
use crate::layouts::model_layout::ModelLayoutInfo;

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Props {
	pub id: String,
	pub inner: Inner,
	pub model_layout_info: ModelLayoutInfo,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type", content = "value")]
pub enum Inner {
	Regressor(RegressorProps),
	BinaryClassifier(BinaryClassifierProps),
	MulticlassClassifier(MulticlassClassifierProps),
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RegressorProps {
	pub baseline_mse: f32,
	pub baseline_rmse: f32,
	pub mse: f32,
	pub rmse: f32,
	pub id: String,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BinaryClassifierProps {
	pub accuracy: f32,
	pub baseline_accuracy: f32,
	pub auc_roc: f32,
	pub id: String,
	pub precision: f32,
	pub recall: f32,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MulticlassClassifierProps {
	pub accuracy: f32,
	pub baseline_accuracy: f32,
	pub class_metrics: Vec<ClassMetrics>,
	pub classes: Vec<String>,
	pub id: String,
}

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ClassMetrics {
	pub precision: f32,
	pub recall: f32,
}
